lib/babe/crypto_test.go

```diff
@@ -2,3 +2,187 @@
 // SPDX-License-Identifier: LGPL-3.0-only
 
 package babe
+
+import (
+	"errors"
+
+	"github.com/ChainSafe/gossamer/lib/crypto/sr25519"
+	"github.com/ChainSafe/gossamer/lib/keystore"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+	"github.com/stretchr/testify/assert"
+
+	"testing"
+)
+
+func TestCalculateThreshold(t *testing.T) {
+	type args struct {
+		C1       uint64
+		C2       uint64
+		numAuths int
+	}
+	tests := []struct {
+		name   string
+		args   args
+		exp    *scale.Uint128
+		expErr error
+	}{
+		{
+			name: "happy path",
+			args: args{
+				C1:       1,
+				C2:       2,
+				numAuths: 3,
+			},
+			exp: &scale.Uint128{Upper: 0x34d00ad6148e1800, Lower: 0x0},
+		},
+		{
+			name: "0 value input",
+			args: args{
+				C1:       0,
+				C2:       0,
+				numAuths: 0,
+			},
+			expErr: ErrThresholdBothZero,
+		},
+		{
+			name: "C1 > C2",
+			args: args{
+				C1:       5,
+				C2:       2,
+				numAuths: 0,
+			},
+			expErr: errors.New("invalid C1/C2: greater than 1"),
+		},
+		{
+			name: "max threshold",
+			args: args{
+				C1:       1,
+				C2:       1,
+				numAuths: 3,
+			},
+			exp: scale.MaxUint128,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			res, err := CalculateThreshold(tt.args.C1, tt.args.C2, tt.args.numAuths)
+			if tt.expErr != nil {
+				assert.EqualError(t, err, tt.expErr.Error())
+			} else {
+				assert.NoError(t, err)
+			}
```

Comment:
Ideally assert the error matches from a programmatic standpoint, using `assert.ErrorIs(t, err, tt.expErr)` where `tt.expErr` is the sentinel error or the wrapped sentinel error.

You can then `assert.EqualError(t, err, tt.expErrMsg)` if you want to assert the message is as expected, this is more from a logging standpoint.

You can compact both with

```go
assert.ErrorIs(t, err, tt.expErr) // takes care of nil as well
if tt.expErr != nil { // only assert message if we expect the error to not be nil so you don't have to set `tt.expErrMsg` to <nil> every time
    assert.EqualError(t, err, tt.expErrMsg)
}
```

Reply:
@qdm12 question: is it better to do error comparisons using strings vs the errors themselves? Also is it worth adding the extra error msg field to all the tests to make this change?

Reply:
I'd say asserting the wrapped error is more important, since some programmatic behavior depending on it (using `errors.Is(err, someSentinelError)` might break without tests detecting it. The string assertion is just for logs really, so we can skip it really.

Although all this is extra test depth, you can leave it as it is now to accelerate merging. At least you're aware for next tests :wink:

Reply:
Okay cool that makes sense, sounds good to me!